### PR TITLE
Use analyzer line number library

### LINKED ISSUE #1938: dartdoc throws exception attempting to find offset inside empty file
We've found it while running `dartdoc` on `native_widgets 1.0.0+4`.

````
Unhandled exception:
RangeError (start): Invalid value: Only valid value is 0: 1
#0      _StringBase.indexOf (dart:core/runtime/libstring_patch.dart:334:7)
#1      String.indexOf (dart:core/runtime/libstring_patch.dart:1024:18)
#2      _createLineNumbersMap (package:dartdoc/src/line_number_cache.dart:30:23)
#3      LineNumberCache.lineAndColumn.<anonymous closure> (package:dartdoc/src/line_number_cache.dart:47:21)
#4      __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.putIfAbsent (dart:collection/runtime/libcompact_hash.dart:281:23)
#5      LineNumberCache.lineAndColumn (package:dartdoc/src/line_number_cache.dart:46:32)
#6      ModelElement.lineAndColumn (package:dartdoc/src/model.dart:3499:40)
#7      new SourceLinker.fromElement (package:dartdoc/src/source_linker.dart:88:27)
#8      ModelElement.sourceHref (package:dartdoc/src/model.dart:3222:25)
````

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #1938, replacing #1939.

This removes the line number cache from dartdoc and also fixes a number of minor problems in addition to #1938:

- All edge cases where we didn't report a location at all are fixed
  - Make synthetic constructors refer to their classes
  - Do something reasonable for the location of an enum's `toString()`
- Make fields and top level variables defer to defined accessors for warnings, rather than the other way around, to make locations for those warnings accurate and distinct
- No longer claim in warnings that documentation is inherited when it isn't

Dartdoc overall is very slightly faster as well (~1%), we were spending a measurable amount time counting lines that the analyzer had already figured out.

### CODE REVIEW

File: lib/src/model.dart [hunk 1]
```diff
@@ -3751,19 +3801,25 @@ abstract class ModelElement extends Canonicalization
...
-  Tuple2<int, int> get lineAndColumn {
-    // TODO(jcollins-g): implement lineAndColumn for explicit fields
-    if (!_isLineNumberComputed) {
-      _lineAndColumn = lineNumberCache.lineAndColumn(
-          element.source.fullName, element.nameOffset);
+  CharacterLocation get characterLocation {
+    if (!_characterLocationIsSet) {
+      LineInfo lineInfo = compilationUnitElement.lineInfo;
+      _characterLocationIsSet = true;
+      assert(element.nameOffset >= 0, 'Invalid location data for element: ${fullyQualifiedName}');
```
Comment (pq) [inline lib/src/model.dart]: I think you can safely remove the braces here and below?  That is just use $fullyQualifiedName...

File: lib/src/model.dart [hunk 2]
```diff
@@ -5273,14 +5319,14 @@ class PackageGraph {
     List<String> messageParts = [warningMessage];
     if (warnable != null) {
       messageParts
-          .add("${warnablePrefix} ${warnableName}: ${warnable.location ?? ''}");
+          .add("${warnablePrefix} ${warnableName}: ${warnable.location}");
```
Comment (pq) [inline lib/src/model.dart]: Again: remove braces?  Up to you obviously but this is at least flutter style...

https://dart-lang.github.io/linter/lints/unnecessary_brace_in_string_interps.html

File: test/model_test.dart [hunk 3]
```diff
@@ -518,7 +518,7 @@ void main() {
       expect(
           packageGraph
               .localPackages.first.defaultCategory.publicLibraries.length,
-          equals(7));
+          equals(8));
```
Comment (pq) [inline test/model_test.dart]: Aside: is `equals` necessary here?  I see the matcher elided below for strings (`expect(package.name, 'test_package');`).  FWIW: I'm on the fence about the short-hand myself.  Needless to say, whatever floats your boat is cool with me! :)

File: lib/src/model.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/model.dart]: yes.  one day I will stop writing bash in dart.

File: lib/src/model.dart (hunk 2 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: test/model_test.dart (hunk 3 above)
Comment (jcollins-g) [reply test/model_test.dart]: leaving the style for integers in place, for now.  a good idea though.

**Pull request merged**